gpt-oss-on-intel-xeon.md

```diff
@@ -0,0 +1,100 @@
+
+<kbd>
+  <img src="assets/gpt-oss-on-intel-xeon/expert_parallelism.png">
+</kbd>
+
+To enable EP on gpt-oss model in transformers, we just need to pass `tp_plan="auto"` when loading the model. We recommend you to use the command `mpirun -np 2 --map-by ppr:1:numa --bind-to numa` to bind the computation of different experts to different NUMA nodes. This ensures that each NUMA node handles its assigned experts independently, leveraging the locality of memory and computation to improve performance and reduce communication overhead.
+
+`mpirun -np 2 --map-by ppr:1:numa --bind-to numa -genv MASTER_ADDR=127.0.0.1 -genv MASTER_PORT=29500 -genv OMP_NUM_THREADS=<cores_per_numa> python tp_hf.py`
+
+```diff
+import os
+import torch
+from transformers import AutoTokenizer, AutoModelForCausalLM
+
+# The mpirun use PMI_RANK and PMI_SIZE as default env to pass rank and world size.
+# We need to set RANK, LOCAL_RANK and WORLD_SIZE which can be recognized by transformers.
+os.environ['RANK'] = str(os.environ.get('PMI_RANK', 0))
+os.environ['LOCAL_RANK'] = str(os.environ.get('PMI_RANK', 0))
+os.environ['WORLD_SIZE'] = str(os.environ.get('PMI_SIZE', 1))
+
+model_id = "lmsys/gpt-oss-20b-bf16"
+# Load model with tp_plan="auto" to enable Expert Parallelism
+model = AutoModelForCausalLM.from_pretrained(
+    model_id,
+    dtype=torch.bfloat16,
++    tp_plan="auto"
+)
+# Prepare input tokens
+tokenizer = AutoTokenizer.from_pretrained(model_id)
+messages = [
+    {"role": "user", "content": "How many rs are in the word 'strawberry'?"},
+]
+inputs = tokenizer.apply_chat_template(
+    messages,
+    add_generation_prompt=True,
+    tokenize=True,
+    return_tensors="pt",
+    return_dict=True,
+).to(model.device)
+
+generated = model.generate(**inputs, max_new_tokens=100)
+print(tokenizer.decode(generated[0][inputs["input_ids"].shape[-1]:]))
+```
+
+
+## Performance Evaluation of Expert Parallelism
+
+To evaluate the performance of Expert Parallelism (EP), we fixed the input and output sequence lengths to 1024 tokens and tested the latency under different batch sizes. The tests were conducted on Intel 6-th Gen Xeon GNR CPUs with and without EP enabled.
+
+The following figures show the performance results for TTFT (Time to First Token) and TPOT (Time per Output Token) under various batch sizes.
+
+<kbd>
+  <img src="assets/gpt-oss-on-intel-xeon/TTFT-gpt-oss.png">
+</kbd>
+
+<kbd>
+  <img src="assets/gpt-oss-on-intel-xeon/TPOT-gpt-oss.png">
+</kbd>
+
+
+## Conclusion
+
+In the TTFT results, we observed that EP could complete the prefill in 1 second for batch size 1. However, as batch size increases, EP becomes slower than non-EP. This is because non-EP utilizes 1 socket with 3 NUMA nodes, while EP involves only 2 NUMA nodes and limits each process to a single NUMA node. Non-EP benefits from more resources, resulting in better TTFT performance at larger batch sizes.
+
```

Comment:
It's improper to talk about 3-NUMA here.

Reply:
done